File: hse_learning/go/contest_Y/num_E.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"unicode"
)

var digitToLetters = []string{
	"",     // 0
	"",     // 1
	"ABC",  // 2
	"DEF",  // 3
	"GHI",  // 4
	"JKL",  // 5
	"MNO",  // 6
	"PQRS", // 7
	"TUV",  // 8
	"WXYZ", // 9
}

func encodeWord(word string) string {
	var encoded strings.Builder
	for _, c := range word {
		c = unicode.ToUpper(c)
		for digit := 2; digit <= 9; digit++ {
			if strings.ContainsRune(digitToLetters[digit], c) {
				pos := strings.IndexRune(digitToLetters[digit], c)
				for i := 0; i <= pos; i++ {
					encoded.WriteByte(byte('0' + digit))
				}
				break
			}
		}
	}
	return encoded.String()
}

func decodeMessage(s string, encodedDictionary map[string]string) string {
	n := len(s)
	dp := make([]bool, n+1)
	dp[0] = true

	maxLen := 0
	for k := range encodedDictionary {
		if len(k) > maxLen {
			maxLen = len(k)
		}
	}

	for i := 0; i < n; i++ {
		if !dp[i] {
			continue
		}
		for length := 1; length <= maxLen && i+length <= n; length++ {
			sub := s[i : i+length]
			if _, ok := encodedDictionary[sub]; ok {
				dp[i+length] = true
			}
		}
	}

	if !dp[n] {
		return ""
	}

	var result []string
	i := n
	for i > 0 {
		found := false
		for length := 1; length <= maxLen && i-length >= 0; length++ {
			sub := s[i-length : i]
			if word, ok := encodedDictionary[sub]; ok && dp[i-length] {
				result = append([]string{word}, result...)
				i -= length
				found = true
				break
			}
		}
		if !found {
			return ""
		}
	}

	return strings.Join(result, " ")
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	s := scanner.Text()

	scanner.Scan()
	var n int
	fmt.Sscanf(scanner.Text(), "%d", &n)

	encodedDictionary := make(map[string]string)
	for i := 0; i < n; i++ {
		scanner.Scan()
		word := scanner.Text()
		encodedWord := encodeWord(word)
		encodedDictionary[encodedWord] = word
	}

	result := decodeMessage(s, encodedDictionary)
	fmt.Println(result)
}

File: hse_learning/cpp_lang/contest_Y/num_E.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>

using namespace std;

vector<string> digitToLetters = {
    "",     // 0
    "",     // 1
    "ABC",  // 2
    "DEF",  // 3
    "GHI",  // 4
    "JKL",  // 5
    "MNO",  // 6
    "PQRS", // 7
    "TUV",  // 8
    "WXYZ"  // 9
};

string encodeWord(const string& word) {
    string encoded;
    for (char c : word) {
        c = toupper(c);
        for (int digit = 2; digit <= 9; ++digit) {
            if (digitToLetters[digit].find(c) != string::npos) {
                int count = digitToLetters[digit].find(c) + 1;
                encoded += string(count, '0' + digit);
                break;
            }
        }
    }
    return encoded;
}

string decodeMessage(const string& s, const unordered_map<string, string>& encodedDictionary) {
    int n = s.length();
    vector<bool> dp(n + 1, false);
    dp[0] = true;

    for (int i = 0; i < n; ++i) {
        if (!dp[i]) continue;
        for (const auto& pair : encodedDictionary) {
            const string& encodedWord = pair.first;
            if (s.substr(i, encodedWord.length()) == encodedWord) {
                dp[i + encodedWord.length()] = true;
            }
        }
    }

    string result;
    int i = n;
    while (i > 0) {
        for (const auto& pair : encodedDictionary) {
            const string& encodedWord = pair.first;
            const string& word = pair.second;
            if (i >= encodedWord.length() && dp[i - encodedWord.length()] &&
                s.substr(i - encodedWord.length(), encodedWord.length()) == encodedWord) {
                result = word + " " + result;
                i -= encodedWord.length();
                break;
            }
        }
    }

    return result;
}

int main() {
    string s;
    cin >> s;

    int n;
    cin >> n;
    unordered_map<string, string> encodedDictionary;
    for (int i = 0; i < n; ++i) {
        string word;
        cin >> word;
        string encodedWord = encodeWord(word);
        encodedDictionary[encodedWord] = word;
    }

    string result = decodeMessage(s, encodedDictionary);

    cout << result << endl;

    return 0;
}